### PR TITLE
New Module : aos_asn_pool as part of network/aos (clean)

### PR DESCRIPTION
**Pull request by dgarros**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
network/aos/aos_asn_pool

##### ANSIBLE VERSION
```
ansible 2.3.0 (aos_asn_pool_clean e06132a1fb) last updated 2017/02/09 14:56:31 (GMT -700)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
I'm working for [Apstra](http://www.apstra.com/), we are making a product to automate Datacenter Networks and we are developed a dozen of Ansible modules and one dynamic inventory to control our product with Ansible.
We have tried to follow Ansible's best practices as much as possible, all our modules are idempotent and support the mode `--check`.

This is the second batch of module that I'm submitting for review, the first `aos_ip_pool` #21044 has been recently approved

@gundalow, please can you take a look and let me know if I need to make some changes 

> I opened a similar PR yesterday for this module but it included the first PR #21044 so I closed it and opened a clean one #21178 . Somehow git was not able to rebase properly  ...

### CODE REVIEW

**Review by jeremyschulman — APPROVED** [review]
shipit

File: lib/ansible/modules/network/aos/aos_asn_pool.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+    # If the module do not exist, return directly
+    if my_pool.exists is False:
+        module.exit_json(changed=False, name=margs['name'], id='', value={})
+
+    ## Check if object is currently in Use or Not
+    # If in Use, return an error
+    if my_pool.value:
+        if my_pool.value['status'] != 'not_in_use':
+            module.fail_json(msg="Unable to delete this Asn Pool is currently in use")
```
Comment (gundalow) [inline lib/ansible/modules/network/aos/aos_asn_pool.py]: Could the error return which Asn Pool is in use?

Also Asn or ASN - This comment applies to a few lines in the modules?

File: lib/ansible/modules/network/aos/aos_asn_pool.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+DOCUMENTATION = '''
+---
+module: aos_asn_pool
+author: Damien Garros (@dgarros)
+version_added: "2.3"
+short_description: Manage AOS ASN Pool
+description:
+  - Apstra AOS ASN Pool module let you manage your ASN Pool easily. You can create
+    create and delete ASN Pool by Name, ID or by using a JSON File. This module
+    is idempotent and support the I(check) mode. It's using the AOS REST API
```
Comment (gundalow) [inline lib/ansible/modules/network/aos/aos_asn_pool.py]: Missing full stop.

File: lib/ansible/modules/network/aos/aos_asn_pool.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+requirements:
+  - "aos-pyez >= 0.6.0"
+options:
+  session:
+    description:
+      - An existing AOS session as obtained by aos_login module
```
Comment (gundalow) [inline lib/ansible/modules/network/aos/aos_asn_pool.py]: Missing full stop.

One all the modules are updates you could raise a single PR to update any reference to another module with the `M(aos_login)`  which will insert a link to the online docs for that module.

e.g.

File: lib/ansible/modules/network/aos/aos_asn_pool.py (hunk 2 above)
Comment (gundalow) [inline lib/ansible/modules/network/aos/aos_asn_pool.py]: support -> supports

File: lib/ansible/modules/network/aos/aos_asn_pool.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+  id:
+    description:
+      - AOS Id of the ASN Pool to manage.
+        Only one of I(name), I(id) or I(content) can be set.
+    required: false
+  content:
+    description:
+      - Datastructure of the ASN Pool to create. The data can be in YAML / JSON or
+        directly a variable. It's the same datastructure that is returned
+        on success in I(value).
```
Comment (gundalow) [inline lib/ansible/modules/network/aos/aos_asn_pool.py]: From the examples is seems like `content` can also be a key when uses with `state: absent`.

File: lib/ansible/modules/network/aos/aos_asn_pool.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+    required: false
+  state:
+    description:
+      - Indicate what is the expected state of the ASN Pool (present or not)
```
Comment (gundalow) [inline lib/ansible/modules/network/aos/aos_asn_pool.py]: Missing full stop.

File: lib/ansible/modules/network/aos/aos_asn_pool.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+    default: present
+    choices: ['present', 'absent']
+    required: false
+  ranges:
+    description:
+      - List of ASNs ranges to add to the ASN Pool
```
Comment (gundalow) [inline lib/ansible/modules/network/aos/aos_asn_pool.py]: Missing full stop.

File: lib/ansible/modules/network/aos/aos_asn_pool.py [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+options:
+  session:
+    description:
+      - An existing AOS session as obtained by aos_login module
+    required: true
+  name:
+    description:
+      - Name of the ASN Pool to manage.
+        Only one of I(name), I(id) or I(content) can be set.
+    required: false
```
Comment (gundalow) [inline lib/ansible/modules/network/aos/aos_asn_pool.py]: You can removed all the `required: false` it's implicit

File: lib/ansible/modules/network/aos/aos_asn_pool.py [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+    return asn_pool.write()
+
+
+def asn_pool_absent(module, aos, my_pool):
+
+    margs = module.params
+
+    # If the module do not exist, return directly
+    if my_pool.exists is False:
+        module.exit_json(changed=False, name=margs['name'], id='', value={})
```
Comment (gundalow) [inline lib/ansible/modules/network/aos/aos_asn_pool.py:190]: Could you please add an example (in reply to his comment) of what this error would look like.

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/aos/aos_asn_pool.py (hunk 8 above)
Comment (dgarros) [reply lib/ansible/modules/network/aos/aos_asn_pool.py:190]: Not sure to understand your question here, if `state: absent` is defined and the pool do not exist, we just return `changed: false` but it's not an error.

Comment (gundalow) [reply lib/ansible/modules/network/aos/aos_asn_pool.py:190]: Ah, yes, I miss read the code. Nothing to change here.

**Comment by dgarros** [comment]
@gundalow, thanks for your review
I made the necessary changes to address your comments, except for one that I'm not sure to understand.
Please let me know if you see something else

**Review by gundalow — APPROVED** [review]

**Comment by dgarros** [comment]
ready_for_review

**Comment by gundalow** [comment]
shipit

**Pull request merged**